File: apps/auth-server/src/main/java/com/example/chat/auth/server/infrastructure/persistence/CredentialEntity.java
package com.example.chat.auth.server.infrastructure.persistence;

import com.example.chat.auth.server.core.domain.Credential;
import com.example.chat.auth.server.core.domain.credential.*;
import jakarta.persistence.*;
import lombok.*;
import lombok.experimental.SuperBuilder;

import java.time.Instant;
import java.util.UUID;

@Entity
@Table(name = "credentials")
@Inheritance(strategy = InheritanceType.SINGLE_TABLE)
@DiscriminatorColumn(name = "credential_type", discriminatorType = DiscriminatorType.STRING)
@Getter
@Setter
@NoArgsConstructor(access = AccessLevel.PROTECTED)
@AllArgsConstructor
@SuperBuilder
public abstract class CredentialEntity {
    @Id
    @Column(name = "id", columnDefinition = "uuid")
    @Builder.Default
    private UUID id = UUID.randomUUID();

    @Column(name = "principal_id", nullable = false)
    private UUID principalId;

    @Column(name = "is_verified", nullable = false)
    @Builder.Default
    private boolean verified = false;

    @Column(name = "created_at", nullable = false, updatable = false)
    @Builder.Default
    private Instant createdAt = Instant.now();

    @Column(name = "credential_type", insertable = false, updatable = false)
    private String type;

    public abstract Credential toDomain();

    public static CredentialEntity fromDomain(UUID principalId, Credential domain) {
        if (domain instanceof PasswordCredential d) {
            return PasswordCredentialEntity.builder()
                    .principalId(principalId)
                    .hashedPassword(d.getHashedPassword() != null ? d.getHashedPassword() : "")
                    .verified(d.isVerified())
                    .build();
        } else if (domain instanceof SocialCredential d) {
            return SocialCredentialEntity.builder()
                    .principalId(principalId)
                    .provider(d.getProvider())
                    .socialUserId(d.getSocialUserId())
                    .email(d.getEmail())
                    .verified(d.isVerified())
                    .build();
        } else if (domain instanceof PasskeyCredential d) {
            return PasskeyCredentialEntity.builder()
                    .principalId(principalId)
                    .credentialId(d.getCredentialId())
                    .publicKey(d.getPublicKey())
                    .authenticatorName(d.getAuthenticatorName())
                    .verified(d.isVerified())
                    .build();
        } else if (domain instanceof OtpCredential d) {
            return OtpCredentialEntity.builder()
                    .principalId(principalId)
                    .code(d.getCode())
                    .deliveryMethod(d.getDeliveryMethod())
                    .verified(false)
                    .build();
        }
        throw new IllegalArgumentException("Unsupported credential type: " + domain.getType());
    }
}

File: apps/auth-server/src/main/kotlin/com/example/chat/auth/server/infrastructure/persistence/CredentialEntity.kt
package com.example.chat.auth.server.infrastructure.persistence

import com.example.chat.auth.server.core.domain.Credential
import com.example.chat.auth.server.core.domain.credential.*
import jakarta.persistence.*
import java.time.Instant
import java.util.*

@Entity
@Table(name = "credentials")
@Inheritance(strategy = InheritanceType.SINGLE_TABLE)
@DiscriminatorColumn(name = "credential_type", discriminatorType = DiscriminatorType.STRING)
abstract class CredentialEntity(
        @Id @Column(name = "id", columnDefinition = "uuid") val id: UUID = UUID.randomUUID(),
        @Column(name = "principal_id", nullable = false) val principalId: UUID,
        @Column(name = "is_verified", nullable = false) val isVerified: Boolean = false,
        @Column(name = "created_at", nullable = false, updatable = false)
        val createdAt: Instant = Instant.now()
) {
    @Column(name = "credential_type", insertable = false, updatable = false)
    val type: String? = null

    abstract fun toDomain(): Credential

    companion object {
        fun fromDomain(principalId: UUID, domain: Credential): CredentialEntity {
            return when (domain) {
                is PasswordCredential ->
                        PasswordCredentialEntity(
                                principalId = principalId,
                                hashedPassword = domain.hashedPassword ?: "",
                                isVerified = domain.isVerified
                        )
                is SocialCredential ->
                        SocialCredentialEntity(
                                principalId = principalId,
                                provider = domain.provider,
                                socialUserId = domain.socialUserId,
                                email = domain.email,
                                isVerified = domain.isVerified
                        )
                is PasskeyCredential ->
                        PasskeyCredentialEntity(
                                principalId = principalId,
                                credentialId = domain.credentialId,
                                publicKey = domain.publicKey,
                                authenticatorName = domain.authenticatorName,
                                isVerified = domain.isVerified
                        )
                is OtpCredential ->
                        OtpCredentialEntity(
                                principalId = principalId,
                                code = domain.code,
                                deliveryMethod = domain.deliveryMethod
                        )
                else ->
                        throw IllegalArgumentException(
                                "Unsupported credential type: ${domain.type}"
                        )
            }
        }
    }
}

@Entity
@DiscriminatorValue("PASSWORD")
class PasswordCredentialEntity(
        principalId: UUID,
        @Column(name = "hashed_password") val hashedPassword: String,
        isVerified: Boolean
) : CredentialEntity(principalId = principalId, isVerified = isVerified) {
    override fun toDomain(): Credential = PasswordCredential(hashedPassword, isVerified)
}

@Entity
@DiscriminatorValue("SOCIAL")
class SocialCredentialEntity(
        principalId: UUID,
        @Column(name = "provider") val provider: String,
        @Column(name = "social_user_id") val socialUserId: String,
        @Column(name = "email") val email: String?,
        isVerified: Boolean
) : CredentialEntity(principalId = principalId, isVerified = isVerified) {
    override fun toDomain(): Credential =
            SocialCredential(provider, socialUserId, email, isVerified)
}

@Entity
@DiscriminatorValue("PASSKEY")
class PasskeyCredentialEntity(
        principalId: UUID,
        @Column(name = "credential_id") val credentialId: String,
        @Column(name = "public_key", length = 1024) val publicKey: String,
        @Column(name = "authenticator_name") val authenticatorName: String?,
        isVerified: Boolean
) : CredentialEntity(principalId = principalId, isVerified = isVerified) {
    override fun toDomain(): Credential =
            PasskeyCredential(credentialId, publicKey, authenticatorName, isVerified)
}

@Entity
@DiscriminatorValue("OTP")
class OtpCredentialEntity(
        principalId: UUID,
        @Column(name = "otp_code") val code: String,
        @Column(name = "delivery_method") val deliveryMethod: String
) : CredentialEntity(principalId = principalId, isVerified = false) {
    override fun toDomain(): Credential = OtpCredential(code, deliveryMethod)
}
